### PR TITLE
SDK-1975 Support for face capture resource

### PR DESCRIPTION
**Pull request by abdalmajeed-yoti**
Add support for:

- Creating a face capture resource
- Retrieving a face capture response
- Uploading a face capture image

### CODE REVIEW

File: src/idv_service/session/create/face_capture/upload.face.capture.image.payload.builder.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @class UploadFaceCaptureImagePayloadBuilder
+ */
+class UploadFaceCaptureImagePayloadBuilder {
+  /**
+   * Sets the content type for uploading a JPEG image
+   *
+   * @return this
+   */
+  forJpegImage() {
```
Comment (liamd-yoti) [inline src/idv_service/session/create/face_capture/upload.face.capture.image.payload.builder.js:16]: Do we really need these `for` methods?

Just wondering if that could become a problem for a consumer.

i.e they use an image like `cat.png` and accidentally use the `.forJpegImage()`. What would happen?

One alternative to these methods is to instead to use https://www.npmjs.com/package/file-type, to detect the file-type from the image buffer itself. That way you are kinda removing the need for these methods. Then it would just be a case of checking if the mimeType is a valid one (image/jpg or image/png).

Comment (abdalmajeed-yoti) [reply src/idv_service/session/create/face_capture/upload.face.capture.image.payload.builder.js:16]: I do agree with you and I wanted to implement an internal file type checker instead of these methods, but I think we need to follow the pattern of the existing same PRs for other languages.

File: package.json [hunk 2]
```diff
@@ -29,6 +29,7 @@
     "coverageDirectory": "./coverage"
   },
   "dependencies": {
+    "form-data": "^4.0.0",
```
Comment (michaelyoti) [inline package.json:32]: It seems like we only use exact versions for packages here. (Maybe we should specify that in an .npmrc)

Comment (abdalmajeed-yoti) [reply package.json]: Yes, I think it would be good to specify that in an `.npmrc`.
But I'm not sure if we should add it in this PR or in a separate one.

File: src/idv_service/session/retrieve/face.capture.resource.response.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+class FaceCaptureResourceResponse extends ResourceResponse {
+  constructor(resource) {
+    super(resource);
+
+    const { image, source } = resource;
+
+    Validation.isPlainObject(image, 'image');
+    this.image = new FaceCaptureImageResponse(image);
+
+    Validation.isPlainObject(source);
```
Comment (michaelyoti) [inline src/idv_service/session/retrieve/face.capture.resource.response.js:17]: Is this missing a second argument?

File: src/idv_service/session/create/face_capture/create.face.capture.resource.payload.builder.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ */
+class CreateFaceCaptureResourcePayloadBuilder {
+  /**
+   * Sets the id of the requirement that the resource will be used to satisfy.
+   *
+   * @param string requirementId
+   * @return CreateFaceCaptureResourcePayloadBuilder
+   */
+  withRequirementId(requirementId) {
+    Validation.isString(requirementId, 'requirement_id');
```
Comment (laurent-yoti) [inline src/idv_service/session/create/face_capture/create.face.capture.resource.payload.builder.js:18]: In every builder that have been added in this PR, the patter for the validation is passing as last parameter the name of the field... so the any error message could include the name of the field.
That is quite useful when validating an object and its internal fields, less so when actually validating some parameters of a method... but for the little benefit it provides, I think we would need to re-use the same name as the parameter (see camel case vs snake case). I have check in the rest of the project, and the rules seems to use the name of the parameter, ie `requirementId` here.

Overall, it would be better to have a nicer message, which let you know that it failed with a parameter {insert name of parameter} for a given method... but maybe we will simply just change the code to improve validation another time so we would not do that.

File: src/request/request.builder.js [hunk 5]
```diff
@@ -176,12 +220,18 @@ class RequestBuilder {
     // Build endpoint and url.
     const endpointPath = `${this.endpoint}?${queryString}`;
 
+    const requestCanSendPayload = yotiCommon.requestCanSendPayload(this.method);
```
Comment (laurent-yoti) [inline src/request/request.builder.js:223]: Just a bit weird to reuse the same name as the method for that const, isn't `canSendPayload` good enough?

File: src/idv_service/session/create/face_capture/create.face.capture.resource.payload.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * @class CreateFaceCaptureResourcePayload
+ */
+class CreateFaceCaptureResourcePayload {
+  /**
+   * @param {string} requirementId
+   */
+  constructor(requirementId) {
+    Validation.isString(requirementId, 'requirementId');
+    this.requirement_id = requirementId;
```
Comment (laurent-yoti) [inline src/idv_service/session/create/face_capture/create.face.capture.resource.payload.js:15]: Keep it camel case.
That is not related directly to the payload, is it ?

Comment (laurent-yoti) [reply src/idv_service/session/create/face_capture/create.face.capture.resource.payload.js:15]: As in, there should be somewhere some conversion... will check more.

File: src/client/idv.client.js [hunk 7]
```diff
@@ -93,6 +93,28 @@ class IDVClient {
     return this.idvService.getSupportedDocuments();
   }
 
+  /**
+   * @param {string} sessionId
+   * @param {CreateFaceCaptureResourcePayload} createFaceCaptureResourcePayload
+   *
+   * @returns {Promise<CreateFaceCaptureResourceResponse>}
+   */
+  createFaceCaptureResource(sessionId, createFaceCaptureResourcePayload) {
```
Comment (michaelyoti) [inline src/client/idv.client.js:102]: Should we add a description for these new methods like the other ones have?

File: examples/idv/index.js [hunk 8]
```diff
@@ -24,6 +24,7 @@ app.use(session({
 const router = express.Router();
 
 router.get('/', controllers.indexController);
+router.get('/fc', controllers.indexControllerWithFaceComparison);
```
Comment (liamd-yoti) [inline examples/idv/index.js:27]: Why `index`?

The other routes are not using that naming minus the actual index with is `/`

File: examples/idv/src/controllers/index.controller.with.face.comparison.js [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+  return idvClient.createSession(sessionSpec);
+}
+
+async function addFaceCaptureResourceToSession(sessionId) {
+  const idvClient = new IDVClient(config.YOTI_CLIENT_SDK_ID, config.YOTI_PEM);
+
+  const sessionConfiguration = await idvClient.getSessionConfiguration(sessionId);
+
+  console.log('>>>>', sessionConfiguration, '<<<< sessionConfiguration');
```
Comment (liamd-yoti) [inline examples/idv/src/controllers/index.controller.with.face.comparison.js:68]: Do we want to keep all these logs around?

File: src/client/idv.client.js [hunk 10]
```diff
@@ -93,6 +93,28 @@ class IDVClient {
...
+
+  /**
+   * @param {string} sessionId
+   * @param {string} resourceId
+   * @param {UploadFaceCaptureImagePayload} uploadFaceCaptureImagePayload
+   *
+   * @returns {Promise<CreateFaceCaptureResourceResponse>}
+   */
+  uploadFaceCaptureImage(sessionId, resourceId, uploadFaceCaptureImagePayload) {
+    // eslint-disable-next-line max-len
```
Comment (liamd-yoti) [inline src/client/idv.client.js:114]: minor:

Just wrap the arguments onto separate lines to avoid the `eslint-disable`

File: src/idv_service/idv.service.js [hunk 11]
```diff
@@ -229,6 +233,66 @@ class IDVService {
     });
   }
 
+  /**
+   * @param {string} sessionId
+   * @param {CreateFaceCaptureResourcePayload} createFaceCaptureResourcePayload
+   *
+   * @returns {Promise} Resolves CreateFaceCaptureResourceResponse
```
Comment (liamd-yoti) [inline src/idv_service/idv.service.js:240]:
```suggestion
   * @returns {Promise<CreateFaceCaptureResourceResponse>}
```

Comment (liamd-yoti) [reply src/idv_service/idv.service.js:240]: Matter of style, elsewhere we had been using the angled brackets syntax to show the resolved type

**Comment by michaelyoti** [comment]
Otherwise LGTM

File: tests/request/payload.spec.js [hunk 12]
```diff
@@ -1,24 +1,86 @@
...
+
+    describe('#getContentType', () => {
+      it('should return the payload content type', () => {
+        const payloadJSON = payload.getContentType();
+        expect(payloadJSON).toBe('application/json');
+      });
+    });
+
+    describe('#getPayloadData', () => {
+      it('should return the payload data as a string', () => {
```
Comment (liamd-yoti) [inline tests/request/payload.spec.js:26]: Just wondering here if you want to assert that what we got back is actually a string that reads how we expect. You have access here to the actual object you passed to the Payload (`const json = { a: '1' }`).

**Review by liamd-yoti — APPROVED** [review]
Just a few comments about style etc, but they are mostly minor things.

Otherwise LG 👍

File: examples/idv/src/controllers/index.controller.with.face.comparison.js (hunk 9 above)
Comment (abdalmajeed-yoti) [reply examples/idv/src/controllers/index.controller.with.face.comparison.js:68]: I think these were adding for testing only and can be removed, right @laurent-yoti ?

Comment (laurent-yoti) [reply examples/idv/src/controllers/index.controller.with.face.comparison.js:68]: yeah, just temporary, let's remove them

File: examples/idv/index.js (hunk 8 above)
Comment (laurent-yoti) [reply examples/idv/index.js:27]: You are right, this was only for checking quickly.
Before merging, we'll check with Dhivya how we want the examples to be.
The current example works with the index directly generating the session... which does not give room for configuring the session... so, yeah, maybe we should refactor so from the index you can configure the session, and therefore having only one route?

File: examples/idv/src/controllers/index.controller.with.face.comparison.js [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+  const faceCaptureResourceRequirements = sessionConfiguration.getCapture().getFaceCaptureResourceRequirements();
+
+  if (faceCaptureResourceRequirements.length) {
+    const [firstRequirement] = faceCaptureResourceRequirements;
+    console.log('>>>>', firstRequirement.getId(), '<<<< firstRequirement.getId()');
```
[the remark this answers is not in the document]
Comment (laurent-yoti) [reply examples/idv/src/controllers/index.controller.with.face.comparison.js:74]: Nope nope nope.

File: tests/idv_service/session/retrieve/allowed.source.response.spec.js [hunk 14]
```diff
@@ -0,0 +1,22 @@
+const AllowedSourceResponse = require('../../../../src/idv_service/session/retrieve/allowed.source.response');
```
Comment (laurent-yoti) [inline tests/idv_service/session/retrieve/allowed.source.response.spec.js:1]: Hum...
We already have a class called `allowed.source.response.js` in `/src/idv_service/session/retrieve/allowed.source.response.js` , so why are we introducing a new one ?

File: tests/idv_service/session/retrieve/allowed.source.response.spec.js [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+
+describe('AllowedSourceResponse', () => {
+  let extendedAllowedSourceResponse;
+
+  beforeEach(() => {
+    class ExtendedAllowedSourceResponse extends AllowedSourceResponse {
+      constructor(type) {
+        super();
+        this.type = type;
+      }
```
Comment (laurent-yoti) [inline tests/idv_service/session/retrieve/allowed.source.response.spec.js:11]: No need to defined the constructor, that will be inherited, and that is exactly what we want to test here, ie, does the implementation in the abstract doing the right thing, so it would be counter productive to have an implementation in the sub class.

File: tests/request/payload.spec.js [hunk 16]
```diff
@@ -1,24 +1,86 @@
...
+        const payloadData = payload.getPayloadData();
+        expect(payloadData).toEqual(JSON.stringify(json));
+      });
+    });
+
+    describe('#getBase64Payload', () => {
+      it('should return the payload data as base64 string', () => {
+        const base64Payload = payload.getBase64Payload();
+        expect(base64Payload).toBe(expectedBase64JsonPayload);
```
Comment (laurent-yoti) [inline tests/request/payload.spec.js:35]: Seems weird to see `expectedBase64JsonPayload` defined above where the actual `json` is defined.
Surely they should be in relation?

File: tests/request/payload.spec.js [hunk 17]
```diff
@@ -1,24 +1,86 @@
...
+      });
     });
   });
 
-  describe('#getBase64Payload', () => {
-    it('should return the base64Payload string', () => {
-      const base64Payload = amlPayload.getBase64Payload();
-      expect(base64Payload).toBe(expectedBase64Payload);
+  describe('when content type is data form', () => {
+    const dataFormFields = {
```
Comment (laurent-yoti) [inline tests/request/payload.spec.js:41]: I struggled to understand what `dataFormFields` was.
So, that is the payload data parameter... not sure what would be a better name...  `payloadData` maybe ?

File: tests/request/payload.spec.js [hunk 18]
```diff
@@ -1,24 +1,86 @@
...
+
+        const formData = new FormData();
+        formData.setBoundary(payload.getRawData().getBoundary());
+        const { name, value } = dataFormFields.getFormDataFields()[0];
+        formData.append(name, value);
+
+        const expectedBase64 = formData.getBuffer().toString('base64');
+
+        expect(base64Payload).toBe(expectedBase64);
+      });
```
Comment (laurent-yoti) [inline tests/request/payload.spec.js:83]: Not sure about these tests.
These are not easy, but here, you are way a bit repeating the implementation...

We could mock `FormData` so we can control what it returns when calling its constructor, so we can make sure that we control what `.getBuffer()` returns, we can also assert that a new FormData is created, and asserting that its `.append()` method is called with the right data ?

File: tests/idv_service/session/retrieve/allowed.source.response.spec.js [hunk 19]
```diff
@@ -0,0 +1,22 @@
+const AllowedSourceResponse = require('../../../../src/idv_service/session/retrieve/allowed.source.response');
+
+describe('AllowedSourceResponse', () => {
+  let extendedAllowedSourceResponse;
+
+  beforeEach(() => {
+    class ExtendedAllowedSourceResponse extends AllowedSourceResponse {
```
[the remark this answers is not in the document]
Comment (abdalmajeed-yoti) [reply tests/idv_service/session/retrieve/allowed.source.response.spec.js:7]: Has been removed and left the other used one.

File: tests/idv_service/session/retrieve/allowed.source.response.spec.js [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+      constructor(type) {
+        super();
+        this.type = type;
+      }
+    }
+
+    extendedAllowedSourceResponse = new ExtendedAllowedSourceResponse('one');
+  });
+
+  describe('#getType', () => {
```
[the remark this answers is not in the document]
Comment (abdalmajeed-yoti) [reply tests/idv_service/session/retrieve/allowed.source.response.spec.js:17]: Has been removed and left the other used one. and that case is tested in the other one.

File: tests/request/payload.spec.js (hunk 18 above)
Comment (abdalmajeed-yoti) [reply tests/request/payload.spec.js]: Yes, but I was thinking of making the test like the ones for json, to have a constant value and tests if the output is the same.
The only issue I had was that the boundary is for the form data is dynamic, so with the new changes I made it static.
